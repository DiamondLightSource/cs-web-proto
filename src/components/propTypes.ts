import PropTypes from "prop-types";

export const StringProp = PropTypes.string.isRequired;
export const StringPropOpt = PropTypes.string;

export const IntProp = PropTypes.number.isRequired;
export const IntPropOpt = PropTypes.number;

export const FloatProp = PropTypes.number.isRequired;
export const FloatPropOpt = PropTypes.number;

export const BoolProp = PropTypes.bool.isRequired;
export const BoolPropOpt = PropTypes.bool;

export const PvProp = PropTypes.string.isRequired;
export const PvPropOpt = PropTypes.string;

export const StringOrNumPropOpt = PropTypes.oneOfType([
  PropTypes.string,
  PropTypes.number
]);
export const StringOrNumProp = StringOrNumPropOpt.isRequired;

// Number of prop types organised into useable sections to form more
// complex units
export const ContainerFeaturesPropType = {
  margin: StringPropOpt,
  padding: StringPropOpt,
  border: StringPropOpt,
  minWidth: StringPropOpt,
  maxWidth: StringPropOpt
};

const OpenWebpagePropType = PropTypes.shape({
  type: StringProp,
  url: StringProp,
  description: StringPropOpt
});

const OpenPagePropType = PropTypes.shape({
  type: StringProp,
  page: StringProp,
  location: StringProp,
  macros: StringProp,
  description: StringPropOpt
});

const ClosePagePropType = PropTypes.shape({
  type: StringProp,
  location: StringProp,
  description: StringPropOpt
});

const WritePvPropType = PropTypes.shape({
  type: StringProp,
  pvName: StringProp,
  value: PropTypes.oneOfType([StringPropOpt, FloatPropOpt]),
  description: StringPropOpt
});

const ActionPropType = PropTypes.oneOfType([
  OpenPagePropType,
  ClosePagePropType,
  WritePvPropType,
  OpenWebpagePropType
]);

export const ActionsPropType = PropTypes.shape({
  executeAsOne: BoolPropOpt,
  actions: PropTypes.arrayOf(ActionPropType).isRequired
});
