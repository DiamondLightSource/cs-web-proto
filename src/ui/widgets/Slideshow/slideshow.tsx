import React, { useState, ReactChildren } from "react";
import PropTypes from "prop-types";
import log from "loglevel";
import { CSSTransition, SwitchTransition } from "react-transition-group";

import classes from "./slideshow.module.css";
import { Widget } from "../widget";
import { WidgetPropType } from "../widgetProps";
import { registerWidget } from "../register";
import {
  ChoicePropOpt,
  InferWidgetProps,
  StringOrNumPropOpt
} from "../propTypes";

import "./styles.css";

const SlideshowProps = {
  children: PropTypes.arrayOf(PropTypes.element),
  overflow: ChoicePropOpt(["scroll", "hidden", "auto", "visible"]),
  maxHeight: StringOrNumPropOpt,
  maxWidth: StringOrNumPropOpt
};

export const SlideshowComponent = (
  props: InferWidgetProps<typeof SlideshowProps>
): JSX.Element => {
  const nextChildIndex = (index: number, length: number): number => {
    if (index + 1 < length) {
      return index + 1;
    } else {
      return 0;
    }
  };
  const previousChildIndex = (index: number, length: number): number => {
    if (index - 1 >= 0) {
      return index - 1;
    } else {
      return length - 1;
    }
  };

  const [childIndex, setChildIndex] = useState(0);

  log.warn(`Slideshow Index: ${childIndex}`);
  //   log.warn(props.children);

  const SwitchableText = (props: { index: number; children: any }) => {
    return (
      <SwitchTransition mode="out-in">
        <CSSTransition
          classNames="alert"
          key={props.index}
          timeout={1000}
          unmountOnExit
          mountOnEnter
        >
          {props.children[props.index]}
        </CSSTransition>
      </SwitchTransition>
    );
  };

  return (
    <div
      className={classes.Slideshow}
      style={{
        maxWidth: props.maxWidth ?? "",
        maxHeight: props.maxHeight ?? ""
      }}
    >
      <button
        style={{
          position: "relative",
          height: "80%",
          width: "10%",
          maxWidth: "3rem",
          maxHeight: "5rem",
          textAlign: "center",
          overflow: "hidden",
          margin: "0.5rem",
          flexGrow: 0
        }}
        onClick={() =>
          setChildIndex(
            previousChildIndex(childIndex, props.children?.length ?? 0)
          )
        }
      >
        ◀
      </button>

      <SwitchableText index={childIndex}>
        <h1>This</h1>
        <h1>Is</h1>
        <h1>Great</h1>
      </SwitchableText>

      <button
        style={{
          position: "relative",
          height: "80%",
          width: "10%",
          maxWidth: "3rem",
          maxHeight: "5rem",
          textAlign: "center",
          overflow: "hidden",
          margin: "0.5rem",
          flexGrow: 0
        }}
        onClick={() =>
          setChildIndex(nextChildIndex(childIndex, props.children?.length ?? 0))
        }
      >
        ▶
      </button>
    </div>
  );
};

const SlideshowWidgetProps = {
  ...SlideshowComponent,
  ...WidgetPropType
};

export const Slideshow = (
  props: InferWidgetProps<typeof SlideshowWidgetProps>
): JSX.Element => <Widget baseWidget={SlideshowComponent} {...props} />;

registerWidget(Slideshow, SlideshowWidgetProps, "slideshow");
