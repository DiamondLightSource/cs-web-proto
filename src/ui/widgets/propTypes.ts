import PropTypes, { InferProps } from "prop-types";

export type ExcludeNulls<T> = {
  [P in keyof T]: Exclude<T[P], null>;
};
export type InferWidgetProps<T> = ExcludeNulls<InferProps<T>>;

export const StringProp = PropTypes.string.isRequired;
export const StringPropOpt = PropTypes.string;

export const IntProp = PropTypes.number.isRequired;
export const IntPropOpt = PropTypes.number;

export const FloatProp = PropTypes.number.isRequired;
export const FloatPropOpt = PropTypes.number;

export const BoolProp = PropTypes.bool.isRequired;
export const BoolPropOpt = PropTypes.bool;

export const PvProp = PropTypes.string.isRequired;
export const PvPropOpt = PropTypes.string;

export const ChildrenProp = PropTypes.node.isRequired;
export const ChildrenPropOpt = PropTypes.node;

export const ObjectProp = PropTypes.object.isRequired;
export const ObjectPropOpt = PropTypes.object;

export const MacrosPropOpt = PropTypes.objectOf(PropTypes.string.isRequired);
export const MacrosProp = PropTypes.objectOf(PropTypes.string.isRequired)
  .isRequired;

export const StringOrNumPropOpt = PropTypes.oneOfType([
  PropTypes.string,
  PropTypes.number
]);
export const StringOrNumProp = StringOrNumPropOpt.isRequired;

export function ChoicePropOpt(
  options: string[]
): PropTypes.Requireable<string> {
  return PropTypes.oneOf(options);
}
export function ChoiceProp(options: string[]): PropTypes.Validator<string> {
  return PropTypes.oneOf(options).isRequired;
}

// Number of prop types organised into useable sections to form more
// complex units
export const ContainerFeaturesPropType = {
  margin: StringPropOpt,
  padding: StringPropOpt,
  border: StringPropOpt,
  minWidth: StringPropOpt,
  maxWidth: StringPropOpt
};

const OpenWebpagePropType = PropTypes.shape({
  type: StringProp,
  openWebpageInfo: PropTypes.shape({
    url: StringProp,
    description: StringPropOpt
  }).isRequired
});

const OpenPagePropType = PropTypes.shape({
  type: StringProp,
  openPageInfo: PropTypes.shape({
    page: StringProp,
    location: StringProp,
    macros: StringProp,
    description: StringPropOpt
  }).isRequired
});

const ClosePagePropType = PropTypes.shape({
  type: StringProp,
  closePageInfo: PropTypes.shape({
    location: StringProp,
    description: StringPropOpt
  }).isRequired
});

const WritePvPropType = PropTypes.shape({
  type: StringProp,
  writePvInfo: PropTypes.shape({
    pvName: StringProp,
    value: PropTypes.oneOfType([StringPropOpt, FloatPropOpt]),
    description: StringPropOpt
  }).isRequired
});

const ActionPropType = PropTypes.oneOfType([
  OpenPagePropType,
  ClosePagePropType,
  WritePvPropType,
  OpenWebpagePropType
]);

export const ActionsPropType = PropTypes.shape({
  executeAsOne: BoolPropOpt,
  actions: PropTypes.arrayOf(ActionPropType).isRequired
});
