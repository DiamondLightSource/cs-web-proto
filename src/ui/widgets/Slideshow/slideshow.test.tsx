import React from "react";
import { mount, ReactWrapper } from "enzyme";
import { create, ReactTestRenderer } from "react-test-renderer";
import { InferWidgetProps } from "../propTypes";

import { SlideshowProps, SlideshowComponent } from "./slideshow";

let snapshot: ReactTestRenderer;
let mounted: ReactWrapper<InferWidgetProps<typeof SlideshowProps>>;

beforeEach((): void => {
  const slideshow = (
    <SlideshowComponent>
      <div id="child-element">Child 1</div>
      <div id="child-element">Child 2</div>
      <div id="child-element">Child 3</div>
    </SlideshowComponent>
  );
  snapshot = create(slideshow);
  mounted = mount(slideshow);
});

describe("<Slideshow />", (): void => {
  test("it matches the snapshot", (): void => {
    expect(snapshot.toJSON()).toMatchSnapshot();
  });

  test("it renders only one child component", () => {
    // console.log(mounted.find("#child-element").debug());
    expect(mounted.find("#child-element")).toHaveLength(1);
  });

  test("it renders the first child on startup", () => {
    expect(mounted.find("#child-element").text()).toEqual("Child 1");
    expect(mounted.find("SwitchableWidget").prop("index")).toEqual(0);
  });

  // Using the index as I know this works and the transition seems to slow
  // down the actual rendering of the new div
  test("next button iterates index", () => {
    mounted.find("#next-button").simulate("click");
    expect(mounted.find("SwitchableWidget").prop("index")).toEqual(1);
  });
  test("next button loops index on overflow", (): void => {
    mounted.find("#next-button").simulate("click");
    expect(mounted.find("SwitchableWidget").prop("index")).toEqual(1);
    mounted.find("#next-button").simulate("click");
    expect(mounted.find("SwitchableWidget").prop("index")).toEqual(2);
    mounted.find("#next-button").simulate("click");
    expect(mounted.find("SwitchableWidget").prop("index")).toEqual(0);
  });
  test("previous button loops around on underflow", (): void => {
    mounted.find("#prev-button").simulate("click");
    expect(mounted.find("SwitchableWidget").prop("index")).toEqual(2);
    mounted.find("#prev-button").simulate("click");
    expect(mounted.find("SwitchableWidget").prop("index")).toEqual(1);
    mounted.find("#prev-button").simulate("click");
    expect(mounted.find("SwitchableWidget").prop("index")).toEqual(0);
  });
});
