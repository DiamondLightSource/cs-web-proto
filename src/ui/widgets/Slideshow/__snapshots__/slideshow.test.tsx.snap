// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Slideshow /> it matches the snapshot 1`] = `
<div
  className="Slideshow"
  style={
    Object {
      "backgroundColor": undefined,
      "maxHeight": "",
      "maxWidth": "",
      "minHeight": "",
    }
  }
>
  <button
    id="prev-button"
    onClick={[Function]}
    style={
      Object {
        "flexGrow": 0,
        "height": "80%",
        "margin": "0.5rem",
        "maxHeight": "5rem",
        "maxWidth": "3rem",
        "overflow": "hidden",
        "position": "relative",
        "textAlign": "center",
        "width": "10%",
      }
    }
  >
    ◀
  </button>
  <div
    style={
      Object {
        "display": "flex",
        "flexGrow": 1,
        "height": "100%",
        "maxHeight": "",
        "overflow": "",
        "position": "relative",
        "width": "80%",
      }
    }
  >
    <div
      id="child-element"
    >
      Child 1
    </div>
  </div>
  <button
    id="next-button"
    onClick={[Function]}
    style={
      Object {
        "flexGrow": 0,
        "height": "80%",
        "margin": "0.5rem",
        "maxHeight": "5rem",
        "maxWidth": "3rem",
        "overflow": "hidden",
        "position": "relative",
        "textAlign": "center",
        "width": "10%",
      }
    }
  >
    ▶
  </button>
</div>
`;
